tap_zendesk/client.py

```diff
@@ -129,8 +121,135 @@ def parse_response(self, response: requests.Response) -> Iterable[dict]:
+
+        if end_date and end_date.tzinfo is None:
+            end_date = end_date.replace(tzinfo=timezone.utc)
+
+        replication_key = self.replication_key
+
+        paginator = self.get_new_paginator()
+        next_page_token = None
+
+        while True:
+            prepared_request = self.prepare_request(context, next_page_token)
+            response = self._request(prepared_request, context)
+            if not response:
+                self.logger.error("Received empty response")
+                break
+
+            records = self.parse_response(response)
+
+            for record in records:
+                if not record:
+                    self.logger.error("Received empty record")
+                    continue
+
+                try:
+                    record_json = json.loads(json.dumps(record))
+                except json.JSONDecodeError as e:
+                    self.logger.error(f"JSONDecodeError encountered while parsing record: {e}")
+                    continue
+
+                record_date_str = record.get(replication_key)
+                if record_date_str:
+                    self.logger.debug(record_date_str)
+                    record_date = datetime.fromisoformat(record_date_str)
+                    if record_date.tzinfo is None:
+                        record_date = record_date.replace(tzinfo=timezone.utc)
+                    if end_date and record_date > end_date:
+                        self.logger.info(
+                            f"Stopping data fetch as record date {record_date} exceeds end date {end_date}")
+                        return
+
+                yield record
+
+            next_page_token = self.get_next_page_token(response)
+            self.logger.debug(f"Next page token: {next_page_token}")
+            if not next_page_token or self.is_end_of_stream(response):
+                break
+
+    def get_next_page_token(self, response):
+        """Get the next page token from the response, if applicable."""
+        if 'after_cursor' in response.json():
+            next_page_token = response.json().get('after_cursor')
+            self.logger.debug(f"Next page token (cursor): {next_page_token}")
+            return next_page_token
+        next_page_token = response.json().get('meta', {}).get('after_cursor')
+        self.logger.debug(f"Next page token (meta): {next_page_token}")
+        return next_page_token
+
+    def is_end_of_stream(self, response):
+        """Determine if the end of stream has been reached."""
+        if 'end_of_stream' in response.json():
+            end_of_stream = response.json().get('end_of_stream')
+            self.logger.debug(f"End of stream: {end_of_stream}")
+            return end_of_stream
+        end_of_stream = not response.json().get('meta', {}).get('after_cursor')
+        self.logger.debug(f"End of stream (meta): {end_of_stream}")
+        return end_of_stream
```

Comment:
I think the sdk can deal with it automatically if you use the `next_page_token_jsonpath` (as I said above if you need to return no page next if the end_date is reached, you can overwrite the `get_new_paginator` but calling the super method first.